library/x509_crt.c

```diff
@@ -1764,6 +1756,104 @@ static int x509_info_subject_alt_name( char **buf, size_t *size,
     return( 0 );
 }
 
+int mbedtls_x509_parse_subject_alternative_name( const mbedtls_x509_crt *crt,
+                                                 mbedtls_x509_subject_alternative_name **san )
+{
+    int ret;
+    const mbedtls_x509_sequence *cur = &crt->subject_alt_names;
+    mbedtls_x509_subject_alternative_name *cur_san = *san, *prev_san = NULL;
+
+    if( crt->ext_types & MBEDTLS_X509_EXT_SUBJECT_ALT_NAME )
+    {
+        if( cur_san != NULL )
+        {
+            return( MBEDTLS_ERR_X509_BAD_INPUT_DATA );
+        }
+
+        while( cur != NULL )
+        {
+            switch( cur->buf.tag &
+                ( MBEDTLS_ASN1_TAG_CLASS_MASK | MBEDTLS_ASN1_TAG_VALUE_MASK ) )
+            {
+                /*
+                 * otherName
+                 */
+                case( MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_X509_SAN_OTHER_NAME ):
+                {
+                    mbedtls_x509_san_other_name other_name;
+
+                    ret = x509_get_other_name( cur, &other_name );
+                    if( ret != 0 )
+                    {
+                        /*
+                         * In case MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE is returned,
+                         * then the "otherName" is of an unsupported type. Ignore.
+                         */
+                        if( ret == MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE )
+                            ret = 0;
+                        continue;
+                    }
+
+                    cur_san = mbedtls_calloc( 1, sizeof( mbedtls_x509_subject_alternative_name ) );
+                    if( cur_san == NULL )
+                        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
+                                MBEDTLS_ERR_ASN1_ALLOC_FAILED );
+
+                    if( prev_san != NULL )
+                        prev_san->next = cur_san;
+
+                    memset( cur_san, 0, sizeof( mbedtls_x509_subject_alternative_name ) );
+                    cur_san->type = MBEDTLS_X509_SAN_OTHER_NAME;
+                    memcpy( &cur_san->san.other_name,
```

Comment:
why do you think? This is a struct, and I would like to copy the internals of the structure `other_name` which is allocated on the stack

Reply:
I think this code does almost the same thing as: `cur_san->san.other_name = other_name`. It should be effectively equivalent to a member-by-member assignment. I can see there's a union inside the struct, and I don't know how that behaves according to the standard, so it may be best to leave the code as it is now. I just prefer to only resort to `memcpy()` if other means don't fit.